services/provider/server/consumer/api_updater.go

```diff
@@ -0,0 +1,86 @@
+package consumer
+
+import (
+	"context"
+	"fmt"
+
+	ocsv1alpha1 "github.com/red-hat-storage/ocs-operator/api/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+type APIUpdater interface {
+	CreateStorageConsumer(sc *ocsv1alpha1.StorageConsumer) error
+	DeleteStorageConsumer(sc *ocsv1alpha1.StorageConsumer) error
+	GetStorageConsumer(name, namespace string) (*ocsv1alpha1.StorageConsumer, error)
+	ListStorageConsumers(namespace string) (*ocsv1alpha1.StorageConsumerList, error)
+	UpdateStorageConsumer(sc *ocsv1alpha1.StorageConsumer) error
+	GetSecret(name, namespace string) (*corev1.Secret, error)
+}
+
+// cAPIUpdater is the consumer API updater
+type cAPIUpdater struct {
+	client client.Client
+}
+
+func NewConsumerAPIUpdater() (APIUpdater, error) {
+	scheme := runtime.NewScheme()
+	err := ocsv1alpha1.AddToScheme(scheme)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add ocsv1alpha1 to scheme. %v", err)
+	}
+	err = corev1.AddToScheme(scheme)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add ocsv1alpha1 to scheme. %v", err)
+	}
+
+	config, err := config.GetConfig()
+	if err != nil {
+		klog.Error(err, "failed to get rest.config")
+		return &cAPIUpdater{}, err
+	}
+	crClient, err := client.New(config, client.Options{Scheme: scheme})
+	if err != nil {
+		klog.Error(err, "failed to create controller-runtime client")
+		return &cAPIUpdater{}, err
+	}
+
+	apiClient := &cAPIUpdater{
+		client: crClient,
+	}
+	return apiClient, nil
+}
+
+func (c cAPIUpdater) CreateStorageConsumer(sc *ocsv1alpha1.StorageConsumer) error {
+	return c.client.Create(context.TODO(), sc)
```

Comment:
Why are we using context.TODO() and not not a single context.Background(), there is not need to switch to a new context for every client invocation.

I would suggest using a single context instance per incoming API request processed by the server

Reply:
Agreed, we need a parent context we pass down to all the callers. Also using a context with cancellation is preferred.

Reply:
passing the parent context provided by the client. It uses cancel with timeout.